adapters/mqtt-vertx-base/src/test/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapterTest.java

```diff
@@ -679,6 +681,49 @@ public void testUploadTelemetryMessageSendsPubAckOnSuccess(final VertxTestContex
                 });
     }
 
+    /**
+     * Verifies that the adapter accepts a command response message with an empty body.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testUploadEmptyCommandResponseSucceeds(final VertxTestContext ctx) {
+
+        // GIVEN an adapter with a command response consumer
+        final CommandResponseSender sender = mock(CommandResponseSender.class);
+        when(sender.sendCommandResponse(any(CommandResponse.class), (SpanContext) any()))
+                .thenReturn(Future.succeededFuture(mock(ProtonDelivery.class)));
+        when(commandConsumerFactory.getCommandResponseSender(anyString(), anyString()))
+                .thenReturn(Future.succeededFuture(sender));
+
+        // WHEN forwarding a command response that has been published
+        final MqttServer server = getMqttServer(false);
+        final AbstractVertxBasedMqttProtocolAdapter<MqttProtocolAdapterProperties> adapter = getAdapter(server);
+
+        final MqttEndpoint endpoint = mockEndpoint();
+        when(endpoint.isConnected()).thenReturn(Boolean.TRUE);
+        final MqttPublishMessage messageFromDevice = mock(MqttPublishMessage.class);
+        when(messageFromDevice.qosLevel()).thenReturn(MqttQoS.AT_MOST_ONCE);
+        when(messageFromDevice.messageId()).thenReturn(5555555);
+        when(messageFromDevice.topicName()).thenReturn("command///res/1010f8ab0b53-bd96-4d99-9d9c-56b868474a6a/200");
+
+        // ... with an empty payload
+        when(messageFromDevice.payload()).thenReturn(null);
+        final ResourceIdentifier address = ResourceIdentifier
+                .fromString("command/my-tenant/4712/res/1010f8ab0b53-bd96-4d99-9d9c-56b868474a6a/200");
+        adapter.uploadCommandResponseMessage(newMqttContext(messageFromDevice, endpoint), address)
+                .setHandler(ctx.completing());
```

Comment:
this  should be `ctx.succeeding(Handler<T>)` and the assertion should be put into the handler. Otherwise the test context will most likely be (successfully) completed before the assertion is executed.
BTW IMHO you should also verify that the command sender has been invoked ...

Reply:
I pushed a fix.